![](images/numpy_matrix.png)

## Learning Objectives

- **Optimize code performance** using vectorization instead of loops
- **Perform vectorized computations** that are 10–100x faster than Python loops  
- **Generate random data** for simulations and statistical analysis   
- **Convert between NumPy and Pandas** for optimal workflow efficiency 
 

In [4]:
import numpy as np
import pandas as pd
import time
import warnings

# Suppress warnings for cleaner output
warnings.filterwarnings('ignore')

##  Vectorization in NumPy

**Vectorization** applies operations to entire arrays simultaneously instead of looping through individual elements. This enables NumPy to leverage highly optimized C/Fortran libraries for dramatic performance gains.

![](https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTQ_9hGutgPqNO9OCLy-vDyJub4iLpgThGZ9A&s)


### Why Vectorization Matters

**Loop Approach (Slow):**
```python
result = []
for i in range(len(array1)):
    for j in range(len(array2)):
        result.append(array1[i] * array2[j] + some_value)
```

**Vectorized Approach (Fast):**
```python
result = array1[:, np.newaxis] * array2 + some_value
```

**Key Benefits:**

| Benefit | Impact |
|---------|--------|
| **Speed** | 10-100x faster than Python loops |
| **Code clarity** | Fewer lines, easier to read |
| **Optimized execution** | Uses compiled C/Fortran libraries |
| **Parallel processing** | Leverages CPU SIMD instructions |
| **Memory efficiency** | Better cache utilization |

###  Performance Comparison: NumPy Vectorization Vs Python Loops

Understanding the performance benefits of vectorization is crucial for writing efficient scientific computing code. Let's compare different scenarios to see when and why vectorization matters.

#### Scenario 1: Basic Mathematical Operations
`np.dot` is a vectorized operation in NumPy that performs matrix multiplication or dot product between arrays. It efficiently computes the element-wise multiplications and then sums them up. To better understand its efficiency, let's first implement the dot product using for loops, and then compare its performance with `np.dot` to see the benefits of vectorization.

In [6]:
# Function to calculate dot product using for loops
def dot_product_loops(arr1, arr2):
    result = 0
    for i in range(len(arr1)):
        result += arr1[i] * arr2[i]
    return result

# Create sample arrays of different sizes for comparison
sizes = [1000, 10000, 100000, 1000000]

print("Performance Comparison: Loops vs NumPy Vectorization")
print("=" * 60)

for size in sizes:
    # Create random arrays
    arr1 = np.random.rand(size)
    arr2 = np.random.rand(size)
    
    # Measure time for the loop-based implementation
    start_time = time.time()
    loop_result = dot_product_loops(arr1, arr2)
    loop_time = time.time() - start_time
    
    # Measure time for np.dot
    start_time = time.time()
    numpy_result = np.dot(arr1, arr2)
    numpy_time = time.time() - start_time
    
    # Calculate speedup
    speedup = loop_time / numpy_time if numpy_time > 0 else float('inf')
    
    print(f"\nArray size: {size:,}")
    print(f"Loop result: {loop_result:.5f}, Time: {loop_time:.5f}s")
    print(f"NumPy result: {numpy_result:.5f}, Time: {numpy_time:.5f}s")
    print(f"Speedup: {speedup:.1f}x faster")
    
    

print(f"\n{'='*60}")
print("Key Observations:")
print("• NumPy becomes dramatically faster as array size increases")
print("• Speedup can reach over 100x for large arrays")
print("• NumPy overhead is minimal for small arrays but pays off quickly")

Performance Comparison: Loops vs NumPy Vectorization

Array size: 1,000
Loop result: 240.51012, Time: 0.00100s
NumPy result: 240.51012, Time: 0.00000s
Speedup: infx faster

Array size: 10,000
Loop result: 2483.13248, Time: 0.00450s
NumPy result: 2483.13248, Time: 0.00000s
Speedup: infx faster

Array size: 100,000
Loop result: 24918.72212, Time: 0.02491s
NumPy result: 24918.72212, Time: 0.00000s
Speedup: infx faster

Array size: 1,000,000
Loop result: 250291.73101, Time: 0.18839s
NumPy result: 250291.73101, Time: 0.00098s
Speedup: 192.8x faster

Key Observations:
• NumPy becomes dramatically faster as array size increases
• Speedup can reach over 100x for large arrays
• NumPy overhead is minimal for small arrays but pays off quickly

Array size: 1,000,000
Loop result: 250291.73101, Time: 0.18839s
NumPy result: 250291.73101, Time: 0.00098s
Speedup: 192.8x faster

Key Observations:
• NumPy becomes dramatically faster as array size increases
• Speedup can reach over 100x for large arrays
•

#### Scenario 2: Complex Mathematical Operations
Let's compare more complex operations that show even greater performance differences:

In [11]:
# Complex mathematical operation: polynomial evaluation
# f(x) = 3x³ + 2x² - 5x + 1

def polynomial_loops(x_values):
    """Evaluate polynomial using loops"""
    results = []
    for x in x_values:
        result = 3*x**3 + 2*x**2 - 5*x + 1
        results.append(result)
    return results

def polynomial_vectorized(x_values):
    """Evaluate polynomial using vectorized operations"""
    return 3*x_values**3 + 2*x_values**2 - 5*x_values + 1

# Test with different sizes
size = 5000000
x_data = np.random.uniform(-10, 10, size)

print("Complex Operations: Polynomial Evaluation")
print("f(x) = 3x³ + 2x² - 5x + 1")
print("-" * 40)

# Loop version
start_time = time.time()
loop_results = polynomial_loops(x_data)
loop_time = time.time() - start_time

# Vectorized version
start_time = time.time()
vector_results = polynomial_vectorized(x_data)
vector_time = time.time() - start_time

speedup = loop_time / vector_time

print(f"Array size: {size:,}")
print(f"Loop time: {loop_time:.4f}s")
print(f"Vectorized time: {vector_time:.4f}s")
print(f"Speedup: {speedup:.1f}x")

# Verify results match
print(f"Results match: {np.allclose(loop_results, vector_results)}")

Complex Operations: Polynomial Evaluation
f(x) = 3x³ + 2x² - 5x + 1
----------------------------------------
Array size: 5,000,000
Loop time: 2.8250s
Vectorized time: 0.2053s
Speedup: 13.8x
Results match: True
Array size: 5,000,000
Loop time: 2.8250s
Vectorized time: 0.2053s
Speedup: 13.8x
Results match: True



The performance comparisons clearly demonstrate that **vectorized NumPy operations significantly outperform equivalent Python loops**


##  Vectorized Operations in NumPy

### Types of Vectorized Operations

NumPy supports several categories of vectorized operations:

####  Element-wise Operations (Universal Functions - ufuncs)
These operations apply a function to each element of an array:

In [5]:
# Basic arithmetic operations (all vectorized)
arr = np.array([1, 4, 9, 16, 25])

print("Original array:", arr)
print("Square root:", np.sqrt(arr))
print("Natural log:", np.log(arr))
print("Sine:", np.sin(arr))
print("Power of 2:", np.power(arr, 2))

# Comparison operations
print("\nComparison operations:")
print("Greater than 10:", arr > 10)
print("Equal to 9:", arr == 9)
print("Between 5 and 20:", (arr >= 5) & (arr <= 20))

Original array: [ 1  4  9 16 25]
Square root: [1. 2. 3. 4. 5.]
Natural log: [0.         1.38629436 2.19722458 2.77258872 3.21887582]
Sine: [ 0.84147098 -0.7568025   0.41211849 -0.28790332 -0.13235175]
Power of 2: [  1  16  81 256 625]

Comparison operations:
Greater than 10: [False False False  True  True]
Equal to 9: [False False  True False False]
Between 5 and 20: [False False  True  True False]


####  Aggregation Operations
These reduce arrays to scalar values or smaller arrays:

In [6]:
# 2D array for aggregation examples
matrix = np.array([[1, 2, 3, 4],
                   [5, 6, 7, 8],
                   [9, 10, 11, 12]])

print("Matrix:")
print(matrix)

# Aggregations across entire array
print(f"\nSum of all elements: {np.sum(matrix)}")
print(f"Mean: {np.mean(matrix)}")
print(f"Standard deviation: {np.std(matrix)}")
print(f"Min: {np.min(matrix)}, Max: {np.max(matrix)}")

# Aggregations along specific axes
print(f"\nSum along axis 0 (columns): {np.sum(matrix, axis=0)}")
print(f"Sum along axis 1 (rows): {np.sum(matrix, axis=1)}")
print(f"Mean along axis 0: {np.mean(matrix, axis=0)}")
print(f"Max along axis 1: {np.max(matrix, axis=1)}")

Matrix:
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]

Sum of all elements: 78
Mean: 6.5
Standard deviation: 3.452052529534663
Min: 1, Max: 12

Sum along axis 0 (columns): [15 18 21 24]
Sum along axis 1 (rows): [10 26 42]
Mean along axis 0: [5. 6. 7. 8.]
Max along axis 1: [ 4  8 12]


#### Boolean Operations and Fancy Indexing
Vectorized selection and filtering operations:

In [7]:
# Boolean indexing - vectorized filtering
data = np.array([1, 15, 8, 23, 4, 16, 42, 3, 19, 7])

# Find elements meeting multiple conditions (vectorized)
mask = (data > 5) & (data < 20)
filtered_data = data[mask]
print(f"Original data: {data}")
print(f"Elements between 5 and 20: {filtered_data}")

# Using np.where for conditional replacement (vectorized)
# Replace values > 15 with -1, others with original value
result = np.where(data > 15, -1, data)
print(f"After conditional replacement: {result}")

# Fancy indexing - select multiple elements at once
indices = [0, 2, 4, 7]
selected = data[indices]
print(f"Elements at indices {indices}: {selected}")

# Count how many elements meet condition (vectorized)
count_large = np.sum(data > 10)
print(f"Count of elements > 10: {count_large}")

Original data: [ 1 15  8 23  4 16 42  3 19  7]
Elements between 5 and 20: [15  8 16 19  7]
After conditional replacement: [ 1 15  8 -1  4 -1 -1  3 -1  7]
Elements at indices [0, 2, 4, 7]: [1 8 4 3]
Count of elements > 10: 5


### Matrix Multiplication in NumPy with Vectorization

Matrix multiplication is one of the most common and computationally intensive operations in numerical computing and deep learning. NumPy offers efficient and highly optimized methods for performing matrix multiplication, which leverage vectorization to handle large matrices quickly and accurately.

**Note**: NumPy matrix operations follow the standard rules of linear algebra, so it's important to ensure that the shapes of the matrices are compatible. If they are not, consider reshaping the matrices before performing multiplication.

There are two common methods for matrix multiplication:

#### Method 1: Matrix Multiplication Using `np.dot()`

In [20]:
# Define two 2D arrays (matrices)
matrix1 = np.array([[1, 2, 3], 
                    [4, 5, 6]])
matrix2 = np.array([[7, 8], 
                    [9, 10], 
                    [11, 12]])

# Matrix multiplication using np.dot
result_dot = np.dot(matrix1, matrix2)
print("Matrix Multiplication using np.dot:\n", result_dot)

# Another way to perform np.dot for matrix multiplication
result_dot2 = matrix1.dot(matrix2)
print("\nMatrix Multiplication using dot method:\n", result_dot2)


Matrix Multiplication using np.dot:
 [[ 58  64]
 [139 154]]

Matrix Multiplication using dot method:
 [[ 58  64]
 [139 154]]


#### Method 2: Matrix Multiplication using `np.matmul()` or `@`

In [21]:
# Matrix multiplication using np.matmul or @ operator
result_matmul = np.matmul(matrix1, matrix2)
result_operator = matrix1 @ matrix2
print("\nMatrix Multiplication using np.matmul:\n", result_matmul)
print("\nMatrix Multiplication using @ operator:\n", result_operator)



Matrix Multiplication using np.matmul:
 [[ 58  64]
 [139 154]]

Matrix Multiplication using @ operator:
 [[ 58  64]
 [139 154]]


**Important:** Note that the `*` operator in NumPy performs element-wise multiplication, not matrix multiplication.

In [ ]:
# Using * operator for element-wise multiplication will cause an error if shapes don't match
# Uncomment the code below to see the error:
# element_wise = matrix1 * matrix2
# print("\nElement-wise Multiplication:\n", element_wise)

In [25]:
# reshape the array for element-wise multiplication
matrix2_reshaped = matrix2.reshape(2, 3)
element_wise = matrix1 * matrix2_reshaped
print("\nElement-wise Multiplication after reshaping:\n", element_wise)


Element-wise Multiplication after reshaping:
 [[ 7 16 27]
 [40 55 72]]


##  Converting Between NumPy Arrays and Pandas DataFrames

**Seamless data conversion** between NumPy and pandas is essential for efficient data science workflows. Each library excels in different areas, and knowing when and how to convert between them maximizes your analytical power.

###  Why Conversion Matters

**NumPy** and **pandas** complement each other perfectly:

| **Aspect** | **NumPy Arrays** | **Pandas DataFrames** |
|------------|------------------|----------------------|
| **Strength** | Mathematical computations | Data manipulation & analysis |
| **Speed** | ⚡ Faster numerical operations | 🔍 Easier data exploration |
| **Memory** | 💾 More memory efficient | 🏷️ Rich metadata (labels, dtypes) |
| **Use Cases** | Linear algebra, statistics, ML algorithms | Data cleaning, grouping, merging |
| **Indexing** | Integer-based only | Label-based + integer-based |
| **Missing Data** | Limited NaN support | 🛡️ Robust NaN handling |



Typical Workflow:

1. **Load data** with pandas (CSV, Excel, databases)
2. **Clean and prepare** with pandas (filtering, grouping)
3. **Convert to NumPy** for intensive computations (ML, statistics)
4. **Convert back to pandas** for results analysis and visualization

###  NumPy Array → Pandas DataFrame

**When to convert**: You have computational results in NumPy arrays and need pandas' data manipulation capabilities.

**Common scenarios**:

-  Adding meaningful labels to numerical results
-  Preparing data for visualization libraries
-  Using pandas' powerful filtering and grouping
-  Exporting to CSV, Excel, or databases

####  Basic Conversion Methods

In [12]:
# 📊 COMPREHENSIVE ARRAY → DATAFRAME CONVERSION
print("🔄 NUMPY TO PANDAS CONVERSION")
print("=" * 40)

# Original NumPy array
scores = np.array([
    [85, 92, 78, 95],
    [88, 76, 91, 82],
    [95, 89, 84, 90]
])

students = ['Alice', 'Bob', 'Carol']
subjects = ['Math', 'Science', 'English', 'History']

print("Original NumPy array:")
print(scores)
print(f"Shape: {scores.shape}")

🔄 NUMPY TO PANDAS CONVERSION
Original NumPy array:
[[85 92 78 95]
 [88 76 91 82]
 [95 89 84 90]]
Shape: (3, 4)


##### Method 1: Basic Conversion

Create a DataFrame directly from a NumPy array using `pd.DataFrame()`. By default, pandas will auto-generate integer labels (0, 1, 2, ...) for both rows and columns.

In [ ]:
# Basic conversion - auto-generated index and column labels
df_basic = pd.DataFrame(scores)
print("\n📋 Basic DataFrame (auto index/columns):")
df_basic


📋 Basic DataFrame (auto index/columns):


,0,1,2,3
0,85,92,78,95
1,88,76,91,82
2,95,89,84,90


Notice that the row indices are `[0, 1, 2]` and column names are `[0, 1, 2, 3]` — these are automatically generated when no labels are specified.

##### Method 2: Custom Row and Column Labels

For better readability and data analysis, you can specify meaningful names for rows (`index`) and columns when creating the DataFrame.

In [15]:
# Custom labels for better data interpretation
df_labeled = pd.DataFrame(scores, index=students, columns=subjects)
print("\n📊 Labeled DataFrame (custom index & columns):")
df_labeled


📊 Labeled DataFrame (custom index & columns):


,Math,Science,English,History
Alice,85,92,78,95
Bob,88,76,91,82
Carol,95,89,84,90


Now the DataFrame has meaningful labels:
- **Row indices**: Student names (Alice, Bob, Carol)
- **Column names**: Subject names (Math, Science, English, History)

This makes the data much more interpretable and easier to work with!

Math       int32
Science    int32
English    int32
History    int32
dtype: object

In [18]:
df_labeled.index

Index(['Alice', 'Bob', 'Carol'], dtype='object')

####  Key Parameters & Options

When converting NumPy arrays to DataFrames, you can customize the output using these parameters:

| **Parameter** | **Purpose** | **Example** | **Default** |
|---------------|-------------|-------------|-------------|
| `data` | NumPy array to convert | `np.array([[1,2],[3,4]])` | Required |
| `columns` | Column labels | `['A', 'B']` | `[0, 1, 2, ...]` |
| `index` | Row labels | `['row1', 'row2']` | `[0, 1, 2, ...]` |


**💡 Best Practices**:

- ✅ **Always specify meaningful column names** for better code readability and maintenance
- ✅ **Use descriptive index names** when your rows represent specific entities (people, dates, etc.)
- ✅ **Check data types** and **index**  after conversion with `df.dtypes` to ensure correct interpretation

In [19]:
df_labeled.dtypes

Math       int32
Science    int32
English    int32
History    int32
dtype: object

In [20]:
df_labeled.index

Index(['Alice', 'Bob', 'Carol'], dtype='object')

In [21]:
# add index name
df_labeled.index.name = 'Student'
df_labeled

,Math,Science,English,History
Student,,,,
Alice,85,92,78,95
Bob,88,76,91,82
Carol,95,89,84,90


###  Pandas DataFrame → NumPy Array

**When to convert**: You need NumPy's computational speed and mathematical functions.

**Common scenarios**:

-  Performance-critical computations (linear algebra, statistics)
-  Integration with scientific computing libraries (SciPy, scikit-learn)
-  Machine learning algorithms that expect NumPy arrays
-  Mathematical operations not available in pandas

####  Conversion Methods & Performance

In [31]:
print(" PANDAS TO NUMPY CONVERSION")
print("=" * 40)

# Start with a DataFrame of tech stock prices
stock_data = pd.DataFrame(
    {
        "AAPL":  [150.0, 152.5, 148.2, 155.1],
        "GOOGL": [2800.0, 2750.0, 2825.0, 2900.0],
        "MSFT":  [300.0, 305.0, 298.5, 310.0],
        "TSLA":  [800.0, 795.0, 820.0, 815.0],
    },
    index=["Day 1", "Day 2", "Day 3", "Day 4"]
)

print("📈 Original DataFrame:")
print(stock_data)
print(f"\nShape: {stock_data.shape}")
print("Dtypes:")
print(stock_data.dtypes)
print(f"Memory usage (deep): {stock_data.memory_usage(deep=True).sum()} bytes")


 PANDAS TO NUMPY CONVERSION
📈 Original DataFrame:
        AAPL   GOOGL   MSFT   TSLA
Day 1  150.0  2800.0  300.0  800.0
Day 2  152.5  2750.0  305.0  795.0
Day 3  148.2  2825.0  298.5  820.0
Day 4  155.1  2900.0  310.0  815.0

Shape: (4, 4)
Dtypes:
AAPL     float64
GOOGL    float64
MSFT     float64
TSLA     float64
dtype: object
Memory usage (deep): 344 bytes


##### Method 1: `.to_numpy()` (Recommended ✅)

The **preferred modern approach** for converting DataFrames to NumPy arrays. This method:
- Converts DataFrame data into a NumPy array
- **Discards** index and column labels (keeps only values)
- Preserves the **same row and column order** as the original DataFrame
- Returns a clean 2D array ready for numerical computations

In [ ]:
# Method 1: .to_numpy() - The modern, recommended approach
print("\n🔢 Method 1 — .to_numpy() (recommended):")
array_modern = stock_data.to_numpy()

print("Converted NumPy array:")
print(array_modern)
print(f"\nShape: {array_modern.shape}")
print(f"Data type: {array_modern.dtype}")
print(f"\n⚠️ Note: Row indices (Day 1, Day 2...) and column names (AAPL, GOOGL...) are lost!")


🔢 Method 1 — .to_numpy() (recommended):
Shape: (4, 4)
Data type: float64


array([[ 150. , 2800. ,  300. ,  800. ],
       [ 152.5, 2750. ,  305. ,  795. ],
       [ 148.2, 2825. ,  298.5,  820. ],
       [ 155.1, 2900. ,  310. ,  815. ]])

**Key Point**: Notice how the resulting array contains **only the numerical values**. All the metadata (index labels like "Day 1", column names like "AAPL") is removed. This is what makes NumPy arrays fast but less descriptive than DataFrames.

##### Method 2: .values (legacy but still works)

In [ ]:
# Method 2: .values - Legacy approach (avoid in new code)
print("\n⚠️  Method 2 — .values (legacy):")
array_legacy = stock_data.values

print("Converted array using .values:")
print(array_legacy)
print(f"\nShape: {array_legacy.shape}")
print(f"Data type: {array_legacy.dtype}")

# Verify both methods produce same result (in this case)
print(f"\nSame result as .to_numpy()? {np.array_equal(array_modern, array_legacy)}")
print("✅ For simple numeric DataFrames, both methods work identically")


⚠️  Method 2 — .values (legacy):
[[ 150.  2800.   300.   800. ]
 [ 152.5 2750.   305.   795. ]
 [ 148.2 2825.   298.5  820. ]
 [ 155.1 2900.   310.   815. ]]
Shape: (4, 4)
Same result as .to_numpy()? True


**Why avoid `.values`?**
- Less clear intent (what does "values" mean?)
- Can behave unpredictably with pandas extension dtypes (nullable integers, strings, etc.)
- Not officially recommended in pandas documentation
- `.to_numpy()` is more explicit and future-proof

##### Method 3: Force a specific dtype (e.g., int32)

In [ ]:
# Method 3: Force a specific dtype during conversion
print("\n🔧 Method 3 — Force data type (int32):")
array_int = stock_data.to_numpy(dtype=np.int32)

print("Array with forced int32 dtype:")
print(array_int)
print(f"\nOriginal dtype: {stock_data.to_numpy().dtype}")
print(f"Forced dtype: {array_int.dtype}")
print(f"\n⚠️ Warning: Float values were truncated (not rounded) to integers!")
print(f"Example: 150.0 → {array_int[0, 0]}, but 152.5 → {array_int[1, 0]} (loses 0.5)")


 Method 3 — Force data type (int32):
[[ 150 2800  300  800]
 [ 152 2750  305  795]
 [ 148 2825  298  820]
 [ 155 2900  310  815]]
Shape: (4, 4)
Data type: int32
Preview (first 2 rows):
[[ 150 2800  300  800]
 [ 152 2750  305  795]]


**When to use `dtype` parameter:**

✅ **Good use cases:**
- Ensuring consistent data types for mathematical operations
- Reducing memory usage (e.g., `float64` → `float32`)
- Preparing data for ML libraries that require specific types
- Converting to integer types when you know there are no decimals

❌ **Be careful:**
- Data loss when converting float to int (truncation, not rounding)
- May raise errors if conversion is impossible (e.g., strings to numbers)
- Check for NaN values before converting to integer types

##### Method 4: Convert specific columns only

In [ ]:
# Method 4: Select and convert specific columns
print("\n🎯 Method 4 — Specific columns only (AAPL, MSFT):")

# Select only AAPL and MSFT columns, then convert
tech_stocks = stock_data[["AAPL", "MSFT"]].to_numpy()

print("Selected columns converted to array:")
print(tech_stocks)
print(f"\nOriginal DataFrame shape: {stock_data.shape} (4 rows × 4 columns)")
print(f"Selected array shape: {tech_stocks.shape} (4 rows × 2 columns)")
print(f"Data type: {tech_stocks.dtype}")
print(f"\n💡 Memory saved: {(stock_data.shape[1] - tech_stocks.shape[1]) / stock_data.shape[1] * 100:.0f}% fewer columns!")


🎯 Method 4 — Specific columns only (AAPL, MSFT):
[[150.  300. ]
 [152.5 305. ]
 [148.2 298.5]
 [155.1 310. ]]
Shape: (4, 2)
Data type: float64


**Why select specific columns?**

- **Performance**: Process only needed data
- **Memory**: Smaller arrays use less RAM
- **Clarity**: Makes your intent explicit
- **Debugging**: Easier to track which data is being used

**Common pattern in data science:**
```python
# Extract features for machine learning
X = df[['feature1', 'feature2', 'feature3']].to_numpy()
y = df['target'].to_numpy()
```

##### Method 2: `.values` (Legacy ⚠️)


An **older method** that still works but is **not recommended** for new code. It behaves similarly to `.to_numpy()` but can produce unexpected results with certain pandas dtypes.
| **Method**                | **Recommendation**           | **Pros**                                              | **Cons**                                              |
|---------------------------|------------------------------|-------------------------------------------------------|-------------------------------------------------------|
| `.to_numpy()`             | ✅ **Use by default**         | Clear intent, stable API, future-proof                | May copy data depending on dtypes/backing storage     |
| `.to_numpy(dtype=...)`    | ✅ **When you need a dtype**  | Explicit, consistent numeric types; easier math       | Casting may be lossy or fail; extra param to manage   |
| `.values`                 | ⚠️ **Legacy / avoid**         | Works on old pandas; quick                           | Ambiguous with mixed/extension dtypes; not future-proof |

**Best practices**
- **Convert only needed columns** to reduce memory and copying.
- **Specify `dtype`** when downstream computations require consistent numeric types.
- **Check for missing values** (`NaN`, `NaT`) before casting to integer dtypes.
- **Prefer `.to_numpy()` over `.values`** for clarity and forward compatibility.
- Be mindful that conversions may **return a copy**, not a view—modify the DataFrame directly if needed.


###  Preserving Metadata During Conversion

**Challenge:** Converting a DataFrame to a NumPy array **drops all metadata**:
- ❌ Column names are lost
- ❌ Index labels are removed  
- ❌ Data type information is simplified or lost
- ❌ Categorical mappings are gone

**Why this matters**: After performing NumPy computations, you often need to convert results back to a DataFrame with the original structure for analysis, visualization, or export.

####  Practical Strategies

##### Strategy 1: Save & Restore Metadata Manually

The most straightforward approach: **save metadata before conversion**, then **restore it after computation**.

In [24]:
print("🔄 STRATEGY 1: Manual Metadata Preservation")
print("=" * 50)

# Create a DataFrame with rich metadata
df_original = pd.DataFrame(
    {
        "Temperature": [72.5, 68.3, 75.1, 71.9],
        "Humidity": [45, 52, 48, 50],
        "WindSpeed": [12.3, 8.7, 15.2, 10.1]
    },
    index=["Monday", "Tuesday", "Wednesday", "Thursday"]
)

print("Original DataFrame:")
print(df_original)
print(f"\nIndex: {df_original.index.tolist()}")
print(f"Columns: {df_original.columns.tolist()}")
print(f"Dtypes:\n{df_original.dtypes}")

# STEP 1: Save metadata BEFORE converting to NumPy
saved_index = df_original.index.copy()
saved_columns = df_original.columns.copy()
saved_dtypes = df_original.dtypes.to_dict()

print("\n✅ Metadata saved!")

# STEP 2: Convert to NumPy for computation
arr = df_original.to_numpy()

print(f"\nNumPy array (metadata lost):")
print(arr)
print(f"Shape: {arr.shape}, dtype: {arr.dtype}")

# STEP 3: Perform computations (example: normalize values)
# Subtract mean and divide by std for each column
arr_normalized = (arr - arr.mean(axis=0)) / arr.std(axis=0)

print(f"\nNormalized array (still no metadata):")
print(arr_normalized)

# STEP 4: Restore metadata when converting back to DataFrame
df_result = pd.DataFrame(
    arr_normalized,
    index=saved_index,
    columns=saved_columns
)

print(f"\n📊 Reconstructed DataFrame with metadata:")
print(df_result)
print(f"\nIndex restored: {df_result.index.tolist()}")
print(f"Columns restored: {df_result.columns.tolist()}")
print(f"Dtypes restored (may differ due to normalization):\n{df_result.dtypes}")

🔄 STRATEGY 1: Manual Metadata Preservation
Original DataFrame:
           Temperature  Humidity  WindSpeed
Monday            72.5        45       12.3
Tuesday           68.3        52        8.7
Wednesday         75.1        48       15.2
Thursday          71.9        50       10.1

Index: ['Monday', 'Tuesday', 'Wednesday', 'Thursday']
Columns: ['Temperature', 'Humidity', 'WindSpeed']
Dtypes:
Temperature    float64
Humidity         int64
WindSpeed      float64
dtype: object

✅ Metadata saved!

NumPy array (metadata lost):
[[72.5 45.  12.3]
 [68.3 52.   8.7]
 [75.1 48.  15.2]
 [71.9 50.  10.1]]
Shape: (4, 3), dtype: float64

Normalized array (still no metadata):
[[ 0.22667166 -1.45010473  0.29531936]
 [-1.50427558  1.25675744 -1.171094  ]
 [ 1.29821043 -0.29002095  1.47659679]
 [-0.02060651  0.48336824 -0.60082214]]

📊 Reconstructed DataFrame with metadata:
           Temperature  Humidity  WindSpeed
Monday        0.226672 -1.450105   0.295319
Tuesday      -1.504276  1.256757  -1.171094

**✅ When to use Strategy 1:**
- Simple DataFrames with basic dtypes (int, float)
- Short-term conversions where you control both sides
- When dtype precision isn't critical

**⚠️ Limitations:**
- Doesn't preserve complex dtypes automatically (Categorical, nullable Int64, datetime with timezone)
- Requires manual tracking of metadata
- Easy to forget to save metadata before conversion

##### Strategy 3: Keep Working in Pandas

**Best practice**: For many operations, you don't need to convert to NumPy at all! Pandas operations are built on NumPy and are often fast enough.

In [ ]:
print("\n💡 STRATEGY 3: Stay in Pandas When Possible")
print("=" * 50)

# Example: Normalize data WITHOUT converting to NumPy
df_data = pd.DataFrame({
    "A": [10, 20, 30, 40],
    "B": [100, 200, 300, 400],
    "C": [1, 2, 3, 4]
}, index=["row1", "row2", "row3", "row4"])

print("Original DataFrame:")
print(df_data)

# Method 1: Using NumPy (metadata handling required)
arr = df_data.to_numpy()
arr_norm = (arr - arr.mean(axis=0)) / arr.std(axis=0)
df_norm_numpy = pd.DataFrame(arr_norm, index=df_data.index, columns=df_data.columns)

print(f"\n📊 Normalized using NumPy (manual metadata restoration):")
print(df_norm_numpy)

# Method 2: Pure Pandas (metadata preserved automatically!)
df_norm_pandas = (df_data - df_data.mean()) / df_data.std()

print(f"\n✨ Normalized using Pandas (metadata automatic):")
print(df_norm_pandas)

# Verify they're the same
print(f"\nResults identical? {np.allclose(df_norm_numpy.values, df_norm_pandas.values)}")

**✅ When to use Strategy 3:**
- Standard mathematical operations (add, multiply, normalize, etc.)
- Statistical computations (mean, std, correlation)
- Element-wise transformations
- Most data analysis tasks

**✅ Pandas operations that preserve metadata:**
- Arithmetic: `df + 10`, `df * 2`, `df / df.mean()`
- Aggregations: `df.sum()`, `df.mean()`, `df.std()`
- Transformations: `df.apply()`, `df.transform()`
- Boolean operations: `df > 0`, `df.between()`

**🚀 Only convert to NumPy when:**
- Using specialized NumPy functions (FFT, linear algebra, advanced random sampling)
- Interfacing with libraries that require NumPy arrays (scikit-learn, TensorFlow)
- Performance-critical nested loops that can be vectorized
- Matrix operations not available in pandas

####  Comparison of Strategies

| **Strategy** | **Preserves Columns** | **Preserves Index** | **Preserves Dtypes** | **Performance** | **Complexity** | **Best For** |
|--------------|:---------------------:|:-------------------:|:--------------------:|:---------------:|:--------------:|--------------|
| **Manual save/restore** | ✅ | ✅ | ⚠️ Partial | ⚡ Fast | 😊 Simple | Most cases |
| **Structured arrays** | ✅ | ⚠️ Optional | ⚠️ Partial | 🐌 Slower | 😰 Complex | Legacy code |
| **Stay in Pandas** | ✅ | ✅ | ✅ Full | ⚡ Fast enough | 😊 Simple | Default choice |

####  Practical Recommendations

**📋 Decision Tree:**

1. **Can you do the operation in pandas?** → ✅ Stay in pandas (Strategy 3)
2. **Need NumPy-specific functions (linear algebra, FFT)?** → Use Strategy 1 (manual save/restore)
3. **Passing to legacy code expecting structured arrays?** → Use Strategy 2 (structured arrays)

**🎯 Best Practice Pattern:**

```python
# 1. Start with pandas DataFrame (has metadata)
df = pd.read_csv('data.csv')

# 2. If you MUST use NumPy:
#    a) Save metadata
saved_index = df.index.copy()
saved_columns = df.columns.copy()

#    b) Convert and compute
arr = df.to_numpy()
result_arr = some_numpy_function(arr)  # Your NumPy computation

#    c) Immediately restore metadata
df_result = pd.DataFrame(result_arr, index=saved_index, columns=saved_columns)

# 3. Continue working with pandas
df_result.to_csv('output.csv')
```

**💡 Key Takeaway**: Metadata loss is the #1 gotcha when converting to NumPy. Always have a plan for restoring it before you convert!

### 📄 Comprehensive Comparison: When to Use What

| **Aspect**           | **Pandas DataFrame**                                  | **NumPy Array**                               | **Best Use**                          |
|----------------------|--------------------------------------------------------|-----------------------------------------------|---------------------------------------|
| **Data labels**      | ✅ Rich metadata (column names, index, dtypes)         | ❌ No labels                                   | **DataFrame** for exploration         |
| **Memory usage**     | ❌ Higher overhead                                     | ✅ Minimal overhead                            | **Array** for large datasets          |
| **Speed**            | ❌ Slower (Python-level checks)                        | ✅ C-level, vectorized performance             | **Array** for heavy compute           |
| **Missing data**     | ✅ Robust (`NaN`, `NaT`, nullable dtypes)              | ⚠️ Limited (mainly float `NaN`)               | **DataFrame** for dirty data          |
| **Indexing**         | ✅ Label & integer (`.loc`, `.iloc`)                   | ❌ Integer/boolean only                        | **DataFrame** for complex queries     |
| **Broadcasting**     | ⚠️ Limited/indirect                                    | ✅ Full NumPy rules                            | **Array** for math ops                |
| **Data types**       | ✅ Mixed types per column                              | ❌ Homogeneous per array                       | **DataFrame** for mixed data          |
| **Plotting**         | ✅ Built-in `.plot()`                                  | ❌ Manual setup                                | **DataFrame** for quick viz           |
| **Linear algebra**   | ❌ Limited                                             | ✅ Rich (NumPy/BLAS/LAPACK)                    | **Array** for math/LA                 |
| **Group operations** | ✅ Powerful `.groupby()`, `.agg()`, `.transform()`     | ❌ Manual                                       | **DataFrame** for aggregations        |


#### 🧭 Recommended Workflow
- 🔍 **Exploratory analysis** → Use **Pandas**
- ⚡ **Heavy computation** → Convert to **NumPy** with `df.to_numpy()`
- 📊 **Results & visualization** → Convert back to **Pandas**
- 💾 **Export & storage** → Use **Pandas** (CSV/Parquet/SQL)

**Flow:** Pandas (explore) → NumPy (compute) → Pandas (analyze/plot) → Pandas (export)

> **Tip:** Convert only the needed columns and set `dtype` in `to_numpy(dtype=...)` when you need consistent numeric types. Handle missing values **before** casting.

##  Random Number Generation in NumPy

NumPy provides a variety of functions for generating random numbers through its numpy.random module. These functions can generate random numbers for different distributions and array shapes, making them essential for simulations, statistical sampling, and machine learning applications.

###  Key Functions for Random Number Generation in NumPy

#### `np.random.rand()`
It generates random numbers from a uniform distribution between 0 and 1.

In [38]:
# Generate a 2x3 array of random values between 0 and 1
rand_array = np.random.rand(2, 3)
print(rand_array)

[[0.39363552 0.47343566 0.85454739]
 [0.34000439 0.86964968 0.08813443]]


#### `np.random.randn()`

It generates random numbers from a standard normal distribution (mean = 0, standard deviation = 1). It is useful for simulating Gaussian distributed data

In [39]:
normal_array = np.random.randn(3, 3)
print(normal_array)

[[ 0.5376299   0.42818625 -0.2799933 ]
 [-1.27903074  0.51529432 -0.83428181]
 [ 2.18409673  0.57050721 -0.5806483 ]]


NumPy's [random](https://numpy.org/doc/stable/reference/random/index.html) module can be used to generate arrays of random numbers from several different probability distributions. For example, a `3x5` array of uniformly distributed random numbers can be generated using the `uniform` function of the `random` module.

#### `np.random.randint()`

It generates random integers within a specific range, you can specify `low` and  `high` values and the shape of the output array

In [40]:
# Generate a 4x4 matrix of random integers between 10 and 20
int_array = np.random.randint(10, 20, (4, 4))
print(int_array)

[[14 17 14 18]
 [15 17 19 17]
 [12 16 16 19]
 [15 16 11 18]]


#### `np.random.choice()`

Randomly selects elements from an input array

In [ ]:
# Select 5 random elements from the array [1, 2, 3, 4, 5] with replacement
choice_array = np.random.choice([1, 2, 3, 4, 5], size=5, replace=True)
print(choice_array)

[4 3 5 2 2]


#### `np.random.normal()`

It generates random numbers from a normal distribution with a specified mean and standard deviation.

In [ ]:
# Generate 6 random numbers from a normal distribution with mean=10 and std=2
custom_normal = np.random.normal(10, 2, size=6)
print(custom_normal)


[ 8.96337061  9.80411528  7.1623284   9.58883996 11.82330829 10.53508984]


#### `np.random.uniform()`

It generates random floating-point numbers between a specified range (low, high) that follow uniform distribution

In [ ]:
# Generate a 1D array of 5 random numbers between -5 and 5
uniform_array = np.random.uniform(-5, 5, size=5)
print(uniform_array)


[-4.76424391 -2.90776829  4.56871733  2.3230159  -4.72083461]


**Example: Creating a 3x5 array of uniform random numbers**

In [ ]:
# Generate a 3x5 array of uniform random numbers between 0 and 1
uniform_2d = np.random.uniform(size=(3, 5))
print(uniform_2d)

array([[0.82164157, 0.10137984, 0.79189132, 0.811572  , 0.482274  ],
       [0.05731031, 0.17458985, 0.70742628, 0.71177794, 0.89835528],
       [0.58268615, 0.12410737, 0.12141491, 0.72113236, 0.18378467]])

For a full list, please check the [official website](https://numpy.org/doc/stable/reference/random/generated/numpy.random.randint.html)

Random numbers can also be generated by Python's built-in [random](https://docs.python.org/3/library/random.html) module. However, it generates one random number at a time, which makes it much slower than NumPy's random module.

## 📝 Key Takeaways

In this notebook, you learned:

1. **Vectorization Fundamentals**
   - How vectorization uses optimized C/Fortran libraries for speed
   - Performance gains of 10-100x over Python loops
   - When to use vectorization vs. loops

2. **Vectorized Operations**
   - Element-wise operations (ufuncs): `np.sqrt()`, `np.log()`, arithmetic operations
   - Aggregations: `np.sum()`, `np.mean()`, `np.std()` with axis control
   - Boolean operations and fancy indexing for filtering

3. **Matrix Operations**
   - Matrix multiplication: `np.dot()`, `np.matmul()`, `@` operator
   - Element-wise vs. matrix multiplication (`*` vs. `@`)
   - Shape compatibility and reshaping

4. **NumPy ↔ Pandas Conversion**
   - When to convert: computation (NumPy) vs. analysis (Pandas)
   - `.to_numpy()` for DataFrame → Array
   - `pd.DataFrame()` for Array → DataFrame
   - Metadata preservation strategies

5. **Random Number Generation**
   - Uniform: `np.random.rand()`, `np.random.uniform()`
   - Normal: `np.random.randn()`, `np.random.normal()`
   - Integers: `np.random.randint()`
   - Sampling: `np.random.choice()`
   - Performance advantages over Python's `random` module


## Independent Study

### Practice Exercise 2: Vectorized Computations with NumPy

This exercise will show vectorized computations with NumPy. Vectorized computations help perform computations more efficiently, and also make the code concise.

**Question:** Read the (1) quantities of roll, bun, cake and bread required by 3 people - Ben, Barbara & Beth, from *food_quantity.csv*, (2) price of these food items in two shops - Target and Kroger, from *price.csv*. Find out which shop should each person go to minimize their expenses.

In [45]:
#Reading the datasets on food quantity and price
import pandas as pd
food_qty = pd.read_csv('./datasets/food_quantity.csv',index_col=0)
price = pd.read_csv('./datasets/price.csv',index_col=0)

In [ ]:
food_qty

,roll,bun,cake,bread
Person,,,,
Ben,6,5,3,1
Barbara,3,6,2,2
Beth,3,4,3,1


In [ ]:
price

,Target,Kroger
Item,,
roll,1.5,1.0
bun,2.0,2.5
cake,5.0,4.5
bread,16.0,17.0


First, let's start from a simple problem. We'll compute the expenses of Ben if he prefers to buy all food items from Target

In [46]:
%%time
# write a for loop to calculate the total cost of Ben's food if he shops at the Target store
total_cost = 0 
for food in food_qty.columns:

    total_cost += food_qty.loc['Ben',food]*price.loc[food,'Target']
total_cost

CPU times: total: 0 ns
Wall time: 1 ms


50.0

In [47]:
%%time
# using numpy 
total_cost = np.sum(food_qty.loc['Ben',]*price.loc[:,'Target'])
total_cost

CPU times: total: 0 ns
Wall time: 0 ns


50.0

Ben will spend $50 if he goes to Target

Now, let's add another layer of complication. We'll compute Ben's expenses for both stores - Target and Kroger

In [48]:
%%time
# using loops to calculate the total cost of food for Ben for all stores
total_cost = {}
for store in price.columns:
    total_cost[store] = 0
    for food in food_qty.columns:
        total_cost[store] += food_qty.loc['Ben',food]*price.loc[food,store]
total_cost

CPU times: total: 15.6 ms
Wall time: 999 μs


{'Target': 50.0, 'Kroger': 49.0}

In [49]:
%%time
# using numpy
total_cost = np.dot(food_qty.loc['Ben',],price.loc[:,:])
total_cost

CPU times: total: 15.6 ms
Wall time: 1 ms


array([50., 49.])

Ben will spend \$50 if he goes to Target, and \$49 if he goes to Kroger. Thus, he should choose Kroger.

Now, let's add the final layer of complication, and solve the problem. We'll compute everyone's expenses for both stores - Target and Kroger

In [50]:
%%timeit
store_expense = pd.DataFrame(0.0, columns=price.columns, index = food_qty.index)
for person in store_expense.index:
    for store in store_expense.columns:
        for food in food_qty.columns:
            store_expense.loc[person, store] += food_qty.loc[person, food]*price.loc[food, store]

store_expense

1.49 ms ± 22.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [51]:
%%timeit
# using matrix multiplication in numpy
pd.DataFrame(np.dot(food_qty.values, price.values), columns=price.columns, index=food_qty.index)

11.3 μs ± 93.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


Based on the above table, Ben should go to Kroger, Barbara to Target and Beth can go to either store.  

As the complexity of operations increases, the number of nested for-loops tends to grow, making the code more cumbersome and difficult to manage. In contrast, leveraging NumPy arrays allows for concise and straightforward implementation, regardless of the complexity. Vectorized computations are not only cleaner but also significantly faster, offering a more efficient solution. To take advantage of this, you first need to convert a pandas DataFrame to a NumPy array for matrix multiplication. Once the computation is complete, you can convert the results back to a pandas DataFrame for further analysis or display

### Practice Exercise 3: Matrix Multiplication for Rating Analysis

Use matrix multiplication to find the average IMDB rating and average Rotten Tomatoes rating for each genre - comedy, action, drama and horror. Use the data: *movies_cleaned.csv*. Which is the most preferred genre for IMDB users, and which is the least preferred genre for Rotten Tomatoes users?

**Hint:** 
1. Create two matrices - one containing the IMDB and Rotten Tomatoes ratings, and the other containing the genre flags (comedy/action/drama/horror). 

2. Multiply the two matrices created in 1.

3. Divide each row/column of the resulting matrix by a vector having the number of ratings in each genre to get the average rating for the genre.

**Solution:**

In [53]:
#| echo: false
data = pd.read_csv('./Datasets/movies_cleaned.csv')
data.head()

,Title,IMDB Rating,Rotten Tomatoes Rating,Running Time min,Release Date,US Gross,Worldwide Gross,Production Budget,comedy,Action,drama,horror
0,Broken Arrow,5.8,55,108,Feb 09 1996,70645997,148345997,65000000,0,1,0,0
1,Brazil,8.0,98,136,Dec 18 1985,9929135,9929135,15000000,1,0,0,0
2,The Cable Guy,5.8,52,95,Jun 14 1996,60240295,102825796,47000000,1,0,0,0
3,Chain Reaction,5.2,13,106,Aug 02 1996,21226204,60209334,55000000,0,1,0,0
4,Clash of the Titans,5.9,65,108,Jun 12 1981,30000000,30000000,15000000,0,1,0,0


In [54]:
#| echo: false
# Getting ratings of all movies
drating = data[['IMDB Rating','Rotten Tomatoes Rating']]
drating_num = drating.to_numpy() #Converting the data to NumPy array
drating_num

array([[ 5.8, 55. ],
       [ 8. , 98. ],
       [ 5.8, 52. ],
       ...,
       [ 7. , 65. ],
       [ 5.7, 26. ],
       [ 6.7, 82. ]])

In [55]:
#| echo: false
# Getting the matrix indicating the genre of all movies
dgenre = data.iloc[:,8:12]
dgenre_num = dgenre.to_numpy() #Converting the data to NumPy array
dgenre_num

array([[0, 1, 0, 0],
       [1, 0, 0, 0],
       [1, 0, 0, 0],
       ...,
       [1, 0, 0, 0],
       [0, 1, 0, 0],
       [0, 1, 0, 0]], dtype=int64)

We'll first find the total IMDB and Rotten tomatoes ratings for all movies of each genre, and then divide them by the number of movies of the corresponding genre to find the average rating for the genre.

For finding the total IMDB and Rotten tomatoes ratings, we'll multiply `drating_num` with `dgenre_num`. However, before multiplying, we'll check if their shapes are compatible for matrix multiplication.

In [56]:

#Shape of drating_num
drating_num.shape

(980, 2)

In [57]:
#Shape of dgenre_num
dgenre_num.shape

(980, 4)

Note that the above shapes are not compatible for matrix multiplication. We'll transpose `dgenre_num` to make the shapes compatible.

In [58]:
#Total IMDB and Rotten tomatoes ratings for each genre
ratings_sum_genre = drating_num.T.dot(dgenre_num)
ratings_sum_genre

array([[ 1785.6,  1673.1,  1630.3,   946.2],
       [14119. , 13725. , 14535. ,  6533. ]])

In [59]:
#Number of movies in the data will be stored in 'rows', and number of columns stored in 'cols'
rows, cols = data.shape

In [60]:
#Getting number of movies in each genre
movies_count_genre = dgenre_num.T.dot(np.ones(rows))
movies_count_genre

array([302., 264., 239., 154.])

In [61]:
#Finding the average IMDB and average Rotten tomatoes ratings for each genre
ratings_sum_genre/movies_count_genre

array([[ 5.91258278,  6.3375    ,  6.82133891,  6.14415584],
       [46.75165563, 51.98863636, 60.81589958, 42.42207792]])

In [62]:
pd.DataFrame(ratings_sum_genre/movies_count_genre,columns = ['comedy','Action','drama','horror'],
             index = ['IMDB Rating','Rotten Tomatoes Rating'])

,comedy,Action,drama,horror
IMDB Rating,5.912583,6.337500,6.821339,6.144156
Rotten Tomatoes Rating,46.751656,51.988636,60.815900,42.422078


IMDB users prefer *drama*, and are amused the least by *comedy* movies, on an average. However, Rotten tomatoes critics would rather watch *comedy* than *horror* movies, on an average.


### Practice Exercise 4: Random Number Generation with NumPy

**Scenario:** Suppose 500 people eat at Food cart 1, and another 500 eat at Food cart 2, everyday.

The waiting time at Food cart 2 has a normal distribution with mean 8 minutes and standard deviation 3 minutes, while the waiting time at Food cart 1 has a uniform distribution with minimum 5 minutes and maximum 25 minutes. 

Simulate a dataset containing waiting times for 500 ppl for 30 days in each of the food joints. Assume that the waiting times are measured simultaneously at a certain time in both places, i.e., the observations are paired.

**On how many days is the average waiting time at Food cart 2 higher than that at Food cart 1?**

**What percentage of times the waiting time at Food cart 2 was higher than the waiting time at Food cart 1?**

Try both approaches: (1) Using loops to generate data, (2) numpy array to generate data. Compare the time taken in both approaches. 

In [2]:
import time as tm

In [3]:
#Method 1: Using loops
start_time = tm.time() #Current system time

#Initializing waiting times for 500 ppl over 30 days with float dtype to avoid warnings
waiting_times_FoodCart1 = pd.DataFrame(0.0, index=range(500), columns=range(30)) #FoodCart1
waiting_times_FoodCart2 = pd.DataFrame(0.0, index=range(500), columns=range(30)) #FoodCart2
import random as rm
for i in range(500):  #Iterating over 500 ppl
    for j in range(30): #Iterating over 30 days
        waiting_times_FoodCart2.iloc[i,j] = rm.gauss(8,3) #Simulating waiting time in FoodCart2 for the ith person on jth day
        waiting_times_FoodCart1.iloc[i,j] = rm.uniform(5,25) #Simulating waiting time in FoodCart1 for the ith person on jth day
time_diff = waiting_times_FoodCart2-waiting_times_FoodCart1

print("On ",sum(time_diff.mean()>0)," days, the average waiting time at FoodCart2 higher than that at FoodCart1")
print("Percentage of times waiting time at FoodCart2 was greater than that at FoodCart1 = ",100*(time_diff>0).sum().sum()/(30*500),"%")
end_time = tm.time() #Current system time
print("Time taken = ", end_time-start_time)

On  0  days, the average waiting time at FoodCart2 higher than that at FoodCart1
Percentage of times waiting time at FoodCart2 was greater than that at FoodCart1 =  15.84 %
Time taken =  0.5275771617889404


In [66]:
#Method 2: Using NumPy arrays
start_time = tm.time()
waiting_time_FoodCart2 = np.random.normal(8,3,size = (500,30)) #Simultaneously generating the waiting times of 500 ppl over 30 days in FoodCart2
waiting_time_FoodCart1 = np.random.uniform(5,25,size = (500,30)) #Simultaneously generating the waiting times of 500 ppl over 30 days in FoodCart1
time_diff = waiting_time_FoodCart2-waiting_time_FoodCart1
print("On ",(time_diff.mean()>0).sum()," days, the average waiting time at FoodCart2 higher than that at FoodCart1")
print("Percentage of times waiting time at FoodCart2 was greater than that at FoodCart1 = ",100*(time_diff>0).sum()/15000,"%")
end_time = tm.time()
print("Time taken = ", end_time-start_time)

On  0  days, the average waiting time at FoodCart2 higher than that at FoodCart1
Percentage of times waiting time at FoodCart2 was greater than that at FoodCart1 =  16.26 %
Time taken =  0.0


The approach with NumPy is much faster than the one with loops.

### Practice Exercise 5: Bootstrapping with NumPy

**Task:** Find the 95% confidence interval of mean profit for 'Action' movies, using Bootstrapping.

Bootstrapping is a non-parametric method for obtaining confidence interval. Use the algorithm below to find the confidence interval:

1. Find the profit for each of the 'Action' movies. Suppose there are *N* such movies. We will have a *Profit* column with *N* values. \
2. Randomly sample *N* values with replacement from the *Profit* column \
3. Find the mean of the *N* values obtained in (b) \
4. Repeat steps (b) and (c) *M=1000* times \
5. The 95% Confidence interval is the range between the 2.5% and 97.5% percentile values of the 1000 means obtained in (c) \
Use the *movies_cleaned.csv* dataset. 

**Solution:**